stablehlo/reference/Ops.cpp

```diff
@@ -550,62 +550,68 @@ Tensor evalDotGeneralOp(const Tensor &lhs, const Tensor &rhs,
-  for (uint64_t i = 0; i < lhsContractingDimensions.size(); ++i) {
-    contractingDimSizes.push_back(
-        lhs.getType().getShape()[lhsContractingDimensions[i]]);
-    totalContractingSize *=
-        lhs.getType().getShape()[lhsContractingDimensions[i]];
-  }
+      rhsResultDims.push_back(i);
 
   for (auto resultIt = result.index_begin(); resultIt != result.index_end();
        ++resultIt) {
-    Index lhsIdx(lhs.getType().getRank());
-    Index rhsIdx(rhs.getType().getRank());
+    // Each result element is computed as dot product of slices of lhs and rhs.
+    // In this implementation, we aren't going to materialize these slices as
+    // standalone tensors, but are going to iterate through lhs and rhs
+    // via lhsIndex and rhsIndex.
+    Index resultIndex = *resultIt;
+    Index lhsIndex(lhs.getType().getRank(), 0);
+    Index rhsIndex(rhs.getType().getRank(), 0);
+
+    // Some pieces of lhsIndex and rhsIndex stay the same during iteration.
+    // These are the indices that correspond to non-contracting dimensions,
+    // and they are initialized here.
     int64_t resultDim = 0;
-
-    // Indices do not change for batching dimensions.
-    for (uint64_t i = 0; i < lhsBatchingDimensions.size(); ++i, ++resultDim) {
-      lhsIdx[lhsBatchingDimensions[i]] = (*resultIt)[resultDim];
-      rhsIdx[rhsBatchingDimensions[i]] = (*resultIt)[resultDim];
+    for (size_t i = 0; i < lhsBatchingDimensions.size(); ++i, ++resultDim) {
+      lhsIndex[lhsBatchingDimensions[i]] = resultIndex[resultDim];
+      rhsIndex[rhsBatchingDimensions[i]] = resultIndex[resultDim];
     }
-
-    // The non-batching, non-contracting dimensions of the operands are copied
-    // over to the result dimensions following the batching dimensions.
-    for (uint64_t i = 0; i < lhsNonBatchingNonContractingDims.size(); i++)
-      lhsIdx[lhsNonBatchingNonContractingDims[i]] = (*resultIt)[resultDim++];
-
-    for (uint64_t i = 0; i < rhsNonBatchingNonContractingDims.size(); i++)
-      rhsIdx[rhsNonBatchingNonContractingDims[i]] = (*resultIt)[resultDim++];
-
-    auto sum = Element(resultType.getElementType(), 0.0);
-    // All pair-wise combination of contracting dimensions needs to be summed.
-    for (int64_t i = 0; i < totalContractingSize; ++i) {
-      sum = sum + lhs.get(lhsIdx) * rhs.get(rhsIdx);
-      if (contractingDimSizes.empty()) continue;
-      for (int64_t j = contractingDimSizes.size() - 1; j >= 0; --j) {
-        lhsIdx[lhsContractingDimensions[j]]++;
-        rhsIdx[rhsContractingDimensions[j]]++;
-        if (lhsIdx[lhsContractingDimensions[j]] != contractingDimSizes[j])
-          break;
-        lhsIdx[lhsContractingDimensions[j]] = 0;
-        rhsIdx[rhsContractingDimensions[j]] = 0;
+    for (size_t i = 0; i < lhsResultDims.size(); ++i, ++resultDim)
+      lhsIndex[lhsResultDims[i]] = resultIndex[resultDim];
+    for (size_t i = 0; i < rhsResultDims.size(); ++i, ++resultDim)
+      rhsIndex[rhsResultDims[i]] = resultIndex[resultDim];
```

Comment:
Brought all lhsIndex/rhsIndex loops together and made them consistent (pre-increment everywhere, `++resultDim` everywhere).